manual/src/tutorials/moduleexamples.etex

```diff
@@ -80,47 +72,39 @@ concerned expression. This can also make the code both easier to read
 (since the code fragment is more self-contained).
 Two constructions are available for this purpose:
 \begin{caml_example}{toplevel}
-let open PrioQueue in
-insert empty 1 "hello";;
+let open Fifo in
+add "hello" empty;;
 \end{caml_example}
 and
 \begin{caml_example}{toplevel}
-PrioQueue.(insert empty 1 "hello");;
+Fifo.(add "hello" empty);;
 \end{caml_example}
 In the second form, when the body of a local open is itself delimited
 by parentheses, braces or bracket, the parentheses of the local open
 can be omitted. For instance,
 \begin{caml_example}{toplevel}
-PrioQueue.[empty] = PrioQueue.([empty]);;
-PrioQueue.[|empty|] = PrioQueue.([|empty|]);;
-PrioQueue.{ contents = empty } = PrioQueue.({ contents = empty });;
-\end{caml_example}
-becomes
-\begin{caml_example}{toplevel}
-PrioQueue.[insert empty 1 "hello"];;
+Fifo.[empty] = Fifo.([empty]);;
+Fifo.[|empty|] = Fifo.([|empty|]);;
+Fifo.{ contents = empty } = Fifo.({ contents = empty });;
 \end{caml_example}
 This second form also works for patterns:
 \begin{caml_example}{toplevel}
 let at_most_one_element x = match x with
-| PrioQueue.( Empty| Node (_,_, Empty,Empty) ) -> true
+| Fifo.{ front = ([] | [_]); rear = [] } -> true
 | _ -> false ;;
 \end{caml_example}
 
 It is also possible to copy the components of a module inside
 another module by using an "include" statement. This can be
 particularly useful to extend existing modules. As an illustration,
 we could add functions that return an optional value rather than
-an exception when the priority queue is empty.
+an exception when the queue is empty.
 \begin{caml_example}{toplevel}
-module PrioQueueOpt =
+module FifoOpt =
 struct
-  include PrioQueue
-
-  let remove_top_opt x =
-    try Some(remove_top x) with Queue_is_empty -> None
-
-  let extract_opt x =
-    try Some(extract x) with Queue_is_empty -> None
+  include Fifo
+  let top_opt q = if is_empty q then None else Some(top q)
+  let pop_opt q = if is_empty q then None else Some(pop q)
```

Comment:
Why not write these with `try ... with`?

Reply:
I think it's easier to understand this way, esp. for readers coming from a Java/C++ background where exceptions are expensive.

Reply:
Makes sense, thanks.